File: backend/src/main/java/jupiterpi/vocabulum/webappserver/sessions/SessionService.java
package jupiterpi.vocabulum.webappserver.sessions;

import jupiterpi.vocabulum.core.sessions.Session;
import jupiterpi.vocabulum.core.users.User;
import jupiterpi.vocabulum.webappserver.db.WebappDatabase;
import jupiterpi.vocabulum.webappserver.db.histories.History;
import jupiterpi.vocabulum.webappserver.db.histories.HistoryItem;
import jupiterpi.vocabulum.webappserver.sessions.cards.CardsSession;
import jupiterpi.vocabulum.webappserver.sessions.chat.ChatSession;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class SessionService {
    private Map<String, WebappSession> sessions = new HashMap<>();

    public String createSession(User user, SessionConfiguration configuration) throws Session.SessionLifecycleException {
        String id = UUID.randomUUID().toString();
        Runnable onComplete = () -> {
            History history = WebappDatabase.get().getHistories().getHistoryOrCreate(user);
            history.addItemAndSave(HistoryItem.createHistoryItem(configuration));
        };
        sessions.put(id, switch (configuration.getMode()) {
            case CHAT -> new ChatSession(configuration.getDirection(), configuration.getSelection(), onComplete);
            case CARDS -> new CardsSession(configuration.getDirection(), configuration.getSelection(), onComplete);
        });
        return id;
    }

    public WebappSession getSession(String id) {
        return sessions.get(id);
    }
}

File: backend/src/main/java/jupiterpi/vocabulum/webappserver/sessions/SessionService.kt
package jupiterpi.vocabulum.webappserver.sessions

import jupiterpi.vocabulum.webappserver.db.models.History
import jupiterpi.vocabulum.webappserver.db.models.User
import jupiterpi.vocabulum.webappserver.sessions.cards.CardsSession
import jupiterpi.vocabulum.webappserver.sessions.chat.ChatSession
import java.util.*

class SessionService {
    private val sessions: MutableMap<String, WebappSession> = HashMap()

    fun createSession(user: User, configuration: SessionConfiguration): String {
        val id = UUID.randomUUID().toString()
        val onComplete = {
            val history = History(user.completeKey, configuration).also { it.save() }
        }
        sessions[id] = when (configuration.mode) {
            Mode.CHAT -> ChatSession(
                configuration.direction,
                configuration.selection,
                onComplete
            )
            Mode.CARDS -> CardsSession(
                configuration.direction,
                configuration.selection,
                onComplete
            )
        }
        return id
    }

    fun getSession(id: String) = sessions[id]
}

interface WebappSession